### DISCUSSION TITLE
Need help with raw SQL querrys that include sub querrys useing .execute()

### DISCUSSION THREAD

**Discussion by ghost** [sqlalchemy/sqlalchemy#9061] [Usage Questions] [answered]
lets say I'm trying to execute this raw SQL query 

**DemoQuery**
```
engine.execute('''
    WITH new_user as (INSERT INTO gs_dev.public.gs_user(username, user_pass, user_email)
    VALUES ('dummy', 'userd', 'email')
    returning user_id)
    INSERT INTO gs_user_state(user_id)
    VALUES ((SELECT user_id from new_user));
    ''')
```
It will run in an PostgreSQL console and INSERT data but when i run it through `engine.execute` the behavior is that of a successful query but no data has been added to the db. if this a bug or have i done something wrong in my usage of execute  

I can get it to run provided I do the query in three distinct SQL statements as it requires a `SELECT` query as well but not if i do it as a individual query SQL using `WITH` and a sub query in plan SQL.

Note: the real query is using prams was just trying to keep the code block a bit cleaner.

**Comment by zzzeek** [comment, marked as answer]
you're relying on a deprecated feature known as "library level autocommit" which parses the SQL to see if COMMIT should be called.  the above statement will not pass this regex so there is no commit.  

run the statement using [modern patterns](https://docs.sqlalchemy.org/en/14/core/connections.html#using-transactions):

```py

with engine.begin() as conn:
    conn.execute('''
    WITH new_user as (INSERT INTO gs_dev.public.gs_user(username, user_pass, user_email)
    VALUES ('dummy', 'userd', 'email')
    returning user_id)
    INSERT INTO gs_user_state(user_id)
    VALUES ((SELECT user_id from new_user));
    ''')

```

**Reply by ghost** [reply]
I'm quite new to sqlalchemy, so I looked in quite a few places to try and find out how to do it correctly but never found any mention of using `with engine.begin()` to execute query's that have sub a query.
What i was searching must have been way off resulting I'm my confusion  :facepalm:  

Thanks a ton!
